blocks/log.py

```diff
@@ -1,274 +1,418 @@
+class TrainingLog(Mapping):
+    r"""The log of training progress.
 
-    To support various backends (such as database tables) the
-    descendants of this class are expected to override the
-    `__getattr__`, `__setattr__`, `__hasattr__` methods.
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Information is represented as nested dictionary
+    of iteration, key values.
 
-    Attributes
-    ----------
-    iterations_done : int
-        The number of iterations done.
-    epochs_done : int
-        The number of epochs done.
-    _epoch_ends : list
-        The numbers of the epochs last iterations.
+    In addition to the set of records of the training progress, a training
+    log has a status object whose attributes describe the current state of
+    training e.g. whether training has started, whether termination has
+    been requested, etc.
 
-    .. todo::
+    Parameters
+    ----------
+    hash_ : str, optional
+        A 12-byte hexidecimal string that identifies this experiment. If
+        given and a database backend is used, it will try to reload the
+        log. If not, a new hash will be created for this log.
+    backend : {'default', 'mongodb', 'sqlite'}, optional
+        The log can be stored in a variety of backends. The `default`
+        backend stores the values in a nested dictionary (a
+        :class:`DefaultOrderedDict` to be precise). This is a simple and
+        fast option, but the log is difficult to access for analysis during
+        training, and can only be saved as a pickled Python object.
+
+        The `mongo` option uses a MongoDB database server to store the
+        logs. The `sqlite` value will use a local SQLite file to store the
+        results. Both of these backends will allow you to analyze results
+        easily even during training, and will allow you to save multiple
+        experiments to the same database.
+    \*\*kwargs
+        Additional keyword arguments are passed to the log's backend e.g.
+        database configuration. See the relevant classes for details.
 
-        We need some notion of an attributes property. Examples of possible
-        properties include a docstring, a priority level to be used by a
-        printing extension to decide whether a certain attribute should be
-        printed or not.
+    Attributes
+    ----------
+    info : dict
+        A dictionary with information (metadata) about this experiment,
+        such as a description of the model, the package versions used, etc.
+        It also contains the unique hash that identifies experiments (e.g.
+        in the database).
+    status : :class:`TrainingStatus`
+        This dictionary contains information on the current status of the
+        training e.g. how many iterations/epochs have been completed,
+        whether training has finished, if errors occurred, etc.
+    hidden : dict
```

Comment:
> What is currently considered private e.g. whether an exception occurred or when an interrupt was intercepted, can actually be very useful if you're writing some sort of monitoring interface which wants to inform the user of the fact that an error occurred.

That's very true, but one can just configure his favourite monitoring front-end to stop ignoring those. That is we should split annotation and its interpretation in exactly the same way we did in `blocks/graph.py`.

In fact (and in this I am basically citing @janchorowski) there is already a good example we can take inspiration from: Python logging packages with its levels, DEBUG, INFO, etc.. 

A random crazy idea: roles of variables monitored by `DataStreamMonitoring` and `TrainingDataMonitoring` can be added as meta-information to the log, which would allow the user to e.g. display only costs in his favourite ncurses based training progress monitor.